patches/server/0962-Plugin-Datapacks.patch

```diff
@@ -0,0 +1,302 @@
++
++        public PackResources create(final String name) {
++            try {
++                final FileSystem fileSystem = FileSystems.newFileSystem(this.pluginProvider().getSource());
++                final Path rootPath = fileSystem.getPath(Objects.requireNonNull(pluginProvider.getMeta().getDatapackConfiguration()).directory());
++                return new PluginDataPackResources(name, rootPath, fileSystem);
++            } catch (final IOException e) {
++                LOGGER.error("Could not load datapack from {}.", this.pluginProvider().getMeta().getName());
++                throw new RuntimeException(e);
++            }
++        }
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/LaunchEntryPointHandler.java b/src/main/java/io/papermc/paper/plugin/entrypoint/LaunchEntryPointHandler.java
+index 61a67971a41527c0e3b614bf48d2bc8eabd443b5..739e406d17297aa5957ead2227e41cc652ba90e8 100644
+--- a/src/main/java/io/papermc/paper/plugin/entrypoint/LaunchEntryPointHandler.java
++++ b/src/main/java/io/papermc/paper/plugin/entrypoint/LaunchEntryPointHandler.java
+@@ -1,9 +1,11 @@
+ package io.papermc.paper.plugin.entrypoint;
+ 
++import io.papermc.paper.datapack.PluginDataPackResources;
+ import io.papermc.paper.plugin.provider.PluginProvider;
+ import io.papermc.paper.plugin.storage.BootstrapProviderStorage;
+ import io.papermc.paper.plugin.storage.ProviderStorage;
+ import io.papermc.paper.plugin.storage.ServerPluginProviderStorage;
++import java.util.List;
+ import org.jetbrains.annotations.ApiStatus;
+ 
+ import java.util.HashMap;
+@@ -27,6 +29,10 @@ public class LaunchEntryPointHandler implements EntrypointHandler {
+         LaunchEntryPointHandler.INSTANCE.enter(Entrypoint.BOOTSTRAPPER);
+     }
+ 
++    public static List<PluginDataPackResources.Loader> getDatapackLoaders() {
++        return ((BootstrapProviderStorage) LaunchEntryPointHandler.INSTANCE.get(Entrypoint.BOOTSTRAPPER)).datapackLoaders;
++    }
++
+     @Override
+     public void enter(Entrypoint<?> entrypoint) {
+         ProviderStorage<?> storage = this.storage.get(entrypoint);
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperDatapackConfiguration.java b/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperDatapackConfiguration.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..5d0c1f2307f2d1f5a96cd1ca1df1a5a70d08e151
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/configuration/PaperDatapackConfiguration.java
+@@ -0,0 +1,9 @@
++package io.papermc.paper.plugin.provider.configuration;
++
++import io.papermc.paper.plugin.DatapackConfiguration;
++import net.kyori.adventure.text.Component;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.checker.nullness.qual.Nullable;
++
++public record PaperDatapackConfiguration(@NonNull String directory, @Nullable Component name) implements DatapackConfiguration {
```

Comment:
idk if we need interfaces for records, that seems a bit weird to me (unless its easier for javadocs this way?)

Reply:
Wdym need? This is the server impl of the interface. We can't just make it a public API record as that means the constructor has to be public.